File: lib/crdt/peer_matrix.rb
module CRDT
  # Keeps track of what each peer knows about each other peer. This structure saves us from having
  # to send around full vector clocks all the time: instead, peers can just send diffs of their
  # vector clock when it is updated, and those diffs are applied to this matrix. It also enables
  # us to use a peer index (a small number) instead of a full 256-bit peer ID, which further
  # reduces message sizes.
  #
  # Each peer has its own mapping from peer IDs to peer indexes (to avoid having to coordinate
  # between peers to agree on a mapping). The only requirement is that for each peer, index 0 is
  # that peer itself. The other indexes are assigned sequentially in arbitrary order by each peer.
  # This matrix keeps track of each peer's assignment of peer indexes to peer IDs.
  class PeerMatrix

    PeerVClockEntry = Struct.new(:peer_id, :peer_index, :op_count)

    # matrix is an array of arrays (i.e. a 2D array).
    # matrix[peer1_index][peer2_index] is a PeerVClockEntry object.
    # Each such object records how many operations peer1 has seen from peer2.
    # peer1_index is according to this peer's local index assignment (see index_by_peer_id);
    # peer2_index is according to peer1's index assignment.
    attr_reader :matrix

    # A hash, where the key is a peer ID (as hex string) and the value is the index that this peer
    # has locally assigned to that peer ID. The indexes must be strictly sequential.
    attr_reader :index_by_peer_id

    # A hash, where the key is a peer ID (as hex string) and the value is a PeerVClockEntry
    # object. This is used to record any operations we see from other peers, so that we can
    # broadcast vector clock diffs to others.
    attr_reader :update_by_peer_id

    def initialize(own_peer_id)
      @matrix = [[PeerVClockEntry.new(own_peer_id, 0, 0)]]
      @index_by_peer_id = {own_peer_id => 0}
      @update_by_peer_id = {}
    end

    def own_peer_id
      @matrix[0][0].peer_id
    end

    # Translates a globally unique peer ID into a local peer index. If the peer ID is not already
    # known, it is added to the matrix and assigned a new index.
    def peer_id_to_index(peer_id)
      index = @index_by_peer_id[peer_id]
      return index if index

      if (@index_by_peer_id.size != @matrix.size) ||
         (@index_by_peer_id.size != @matrix[0].size) ||
          @matrix[0].any? {|entry| entry.peer_id == peer_id }
        raise 'Mismatch between vector clock and peer list'
      end

      index = @index_by_peer_id.size
      @index_by_peer_id[peer_id] = index
      @matrix[0][index] = PeerVClockEntry.new(peer_id, index, 0)
      @matrix[index] = [PeerVClockEntry.new(peer_id, 0, 0)]
      @update_by_peer_id[peer_id] = PeerVClockEntry.new(peer_id, index, 0)
      index
    end

    # For an incoming message from the given peer ID, check that the peer is set up in our local
    # state, and that the operation counter lines up with what we were expecting.
    def process_incoming_op_count(origin_peer_id, origin_op_count)
      origin_index = peer_id_to_index(origin_peer_id)

      # We normally expect the opCount for a peer to be monotonically increasing. However, there's
      # a possible scenario in which a peer sends some messages and then crashes before writing its
      # state to stable storage, so when it comes back up, it reverts back to a lower opCount.
      # We should detect when this happens, and replay the lost messages from another peer.
      entry = @matrix[0][origin_index]
      raise "peerID mismatch: #{entry.peer_id} != #{origin_peer_id}" if entry.peer_id != origin_peer_id
      raise "opCount for #{origin_peer_id} went backwards"  if entry.op_count > origin_op_count
      raise "opCount for #{origin_peer_id} jumped forwards" if entry.op_count < origin_op_count
    end

    # Processes a clock update from a remote peer and applies it to the local state.
    # This update indicates that +origin_peer_id+ has received +op_count+ operations from the peer
    # with ID +subject_peer_id+. Moreover, +origin_peer_id+ has assigned index
    # +subject_peer_index+ to that peer. +subject_peer_id+ may be left nil if +origin_peer_id+ has
    # already broadcast its mapping from +subject_peer_id+ to +subject_peer_index+ previously.
    def clock_update(origin_peer_id, subject_peer_id, subject_peer_index, op_count)
      vclock = @matrix[@index_by_peer_id[origin_peer_id]]
      entry = vclock[subject_peer_index]

      if entry.nil?
        raise 'Non-consecutive peer index assignment' if subject_peer_index != vclock.size
        raise 'New peer index assignment without ID' if subject_peer_id.nil?
        entry = PeerVClockEntry.new(subject_peer_id, subject_peer_index, op_count)
        vclock[subject_peer_index] = entry
      else
        entry.op_count = op_count
      end
    end

    # Increments the operation counter for the local peer, indicates that an operation has been
    # performed locally.
    def local_operation
      @matrix[0][0].op_count += 1
    end

    # Increments the operation counter for a particular peer, indicating that we have processed an
    # operation that originated on that peer. In other words, this moves the vector clock forward.
    def increment_op_count(origin_peer_id)
      origin_index = peer_id_to_index(origin_peer_id)

      local_vclock_entry = @matrix[0][origin_index]
      local_vclock_entry.op_count += 1

      remote_vclock_entry = @matrix[origin_index][0]
      remote_vclock_entry.op_count += 1

      @update_by_peer_id[origin_peer_id] ||= PeerVClockEntry.new(nil, origin_index, 0)
      @update_by_peer_id[origin_peer_id].op_count = local_vclock_entry.op_count
    end
  end
end

File: java/src/main/java/org/trvedata/PeerMatrix.java
package org.trvedata;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class PeerMatrix {
	
	private static class PeerVClockList implements Iterable<PeerVClockEntry> {
		List<PeerVClockEntry> vClocks;

		public PeerVClockList(List<PeerVClockEntry> vClocks) {
			this.vClocks = new ArrayList<PeerVClockEntry>(vClocks);
		}

		public PeerVClockEntry getClockEntry(PeerIndex i) {
			return i.idx < vClocks.size() ? vClocks.get(i.idx) : null;
		}

		public int size() {
			return this.vClocks.size();
		}

		public boolean hasPeerId(String peerId) {
			for (PeerVClockEntry entry : this.vClocks)
				if (entry.getPeerId().equals(peerId))
					return true;
			return false;
		}

		public PeerVClockEntry setClockEntry(PeerIndex subjectPeerIndex, PeerVClockEntry entry) {
			while (this.vClocks.size() <= subjectPeerIndex.idx)
				this.vClocks.add(null);
			this.vClocks.set(subjectPeerIndex.idx, entry);
			return entry;
		}

		@Override
		public Iterator<PeerVClockEntry> iterator() {
			return this.vClocks.iterator();
		}

		public void add(PeerVClockEntry peerVClockEntry) {
			this.vClocks.add(peerVClockEntry);
		}

		@Override
		public String toString() {
			return "MatrixEntry [vClocks=" + vClocks + "]";
		}
	}
	
	private static final PeerIndex PEER_INDEX_LOCAL = new PeerIndex(0);
	
	private List<PeerVClockList> vectorClockMatrix;
	private HashMap<String, PeerIndex> indexByPeerId;
	private LocalClockUpdate localClockUpdate;

	public PeerMatrix(String ownPeerId) {
		// matrix.get(peer1Index).get(peer2Index) records how many operations peer1 has seen from peer2. peer1Index is
		// according to this peer's local index assignment (see indexByPeerId); peer2Index is according to peer1's index
		// assignment.
		this.vectorClockMatrix = new ArrayList<PeerVClockList>();
		this.vectorClockMatrix.add(new PeerVClockList(Arrays.asList(new PeerVClockEntry(ownPeerId, PEER_INDEX_LOCAL, 0))));
		// Key: peer ID (as hex string), Value: the index that this peer has locally assigned to that peer ID. The
		// indexes must be strictly sequential.
		this.indexByPeerId = new HashMap<String, PeerIndex>();
		this.indexByPeerId.put(ownPeerId, PEER_INDEX_LOCAL);
		// used to record any operations we see from other peers, so that we can broadcast vector clock diffs to others.
		this.localClockUpdate = new LocalClockUpdate();
	}

	/**
	 * Returns the peer ID (globally unique hex string) for the local device.
	 */
	public String ownPeerId() {
		return getPeerVClockList(PEER_INDEX_LOCAL).getClockEntry(PEER_INDEX_LOCAL).getPeerId();
	}
	
	protected PeerVClockList getPeerVClockList(PeerIndex peerIndex) {
		return this.vectorClockMatrix.get(peerIndex.idx);
	}
	
	protected void setPeerVClockList(PeerIndex peerIndex, PeerVClockList peerVClockList) {
		this.vectorClockMatrix.set(peerIndex.idx, peerVClockList);
	}
	
	private void addPeerVClockList(PeerVClockList peerVClockList) {
		this.vectorClockMatrix.add(peerVClockList);
	}

	/**
	 * When we get a message from originPeerId, it may refer to another peer by an integer index remotePeerIndex. This
	 * method translates remotePeerIndex (which is meaningful only in the context of messages from originPeerId) to the
	 * corresponding peer Id (a hex string that is globally unique).
	 */
	public String remoteIndexToPeerId(String originPeerId, PeerIndex remotePeerIndex) {
		PeerVClockEntry entry = getPeerVClockList(this.peerIdToIndex(originPeerId)).getClockEntry(remotePeerIndex);
		if (entry == null)
			throw new RuntimeException("remoteIndexToPeerId: No peer Id for index " + remotePeerIndex);
		return entry.getPeerId();
	}

	/**
	 * Translates a globally unique peer ID into a local peer index. If the peer ID is not already known, it is added to
	 * the matrix and assigned a new index.
	 */
	public PeerIndex peerIdToIndex(String peerId) {
		PeerIndex index = this.indexByPeerId.get(peerId);
		if (index != null)
			return index;
		index = new PeerIndex(this.indexByPeerId.size());
		
		assert index.idx == this.vectorClockMatrix.size();
		assert index.idx == this.vectorClockMatrix.get(PEER_INDEX_LOCAL.idx).size();
		assert !this.vectorClockMatrix.get(PEER_INDEX_LOCAL.idx).hasPeerId(peerId);
		
		this.indexByPeerId.put(peerId, index);
		getPeerVClockList(PEER_INDEX_LOCAL).add(new PeerVClockEntry(peerId, index, 0));
		addPeerVClockList(new PeerVClockList(Arrays.asList(new PeerVClockEntry(peerId, PEER_INDEX_LOCAL, 0))));
		this.localClockUpdate.addPeer(peerId, index);
		return index;
	}

	/*
	 * Indicates that the peer originPeerId has assigned an index of subjectPeerIndex to the peer subjectPeerId. Calling
	 * this method registers the mapping, so that subsequent calls to remoteIndexToPeerId can resolve the index. Returns
	 * the appropriate PeerVClockEntry.
	 */
	protected PeerVClockEntry peerIndexMapping(String originPeerId, String subjectPeerId, PeerIndex subjectPeerIndex) {
		PeerVClockList vclocks = getPeerVClockList(this.peerIdToIndex(originPeerId));
		PeerVClockEntry entry = vclocks.getClockEntry(subjectPeerIndex);
		if (entry != null) {
			if (subjectPeerId != null && subjectPeerId != entry.getPeerId())
				throw new RuntimeException("peerIndexMapping: Contradictory peer index assignment: " + subjectPeerId + " != " + entry.getPeerId());
			return entry;
		} else if (subjectPeerIndex.idx != vclocks.size()) {
			throw new RuntimeException("peerIndexMapping: Non-consecutive peer index assignment: " + subjectPeerIndex + " != " + vclocks.size());
		} else if (subjectPeerId == null) {
			throw new RuntimeException("peerIndexMapping: New peer index assignment without ID");
		} else {
			return vclocks.setClockEntry(subjectPeerIndex, new PeerVClockEntry(subjectPeerId, subjectPeerIndex, 0));
		}
	}

	/*
	 * Processes a clock update from a remote peer and applies it to the local state. The update indicates that
	 * originPeerId has received various operations from other peers, and also documents which peer indexes originPeerId
	 * has assigned to those peers.
	 */
	protected void applyClockUpdate(String originPeerId, ClockUpdate update) {
		for (PeerVClockEntry newEntry : update.entries()) {
			PeerVClockEntry oldEntry = this.peerIndexMapping(originPeerId, newEntry.getPeerId(), newEntry.getPeerIndex());
			if (oldEntry.getMsgCount() > newEntry.getMsgCount())
				throw new RuntimeException("applyClockUpdate: Clock update went backwards: " + oldEntry.getMsgCount() + " > " + newEntry.getMsgCount());
			oldEntry.setMsgCount(newEntry.getMsgCount());
		}
	}

	/*
	 * Increments the message counter for the local peer, indicating that a message has been broadcast to other peers.
	 */
	protected long incrementMsgCount() {
		return getPeerVClockList(PEER_INDEX_LOCAL).getClockEntry(PEER_INDEX_LOCAL).incrementMsgCount();
	}

	/*
	 * Increments the message counter for a particular peer, indicating that we have processed a message that originated
	 * on that peer. In other words, this moves the vector clock forward.
	 */
	protected void processedIncomingMsg(String originPeerId, long msgCount) {
		PeerIndex originIndex = this.peerIdToIndex(originPeerId);
		PeerVClockEntry localEntry = getPeerVClockList(PEER_INDEX_LOCAL).getClockEntry(originIndex);
		PeerVClockEntry remoteEntry = getPeerVClockList(originIndex).getClockEntry(PEER_INDEX_LOCAL);

		// We normally expect the msgCount for a peer to be monotonically increasing. However, there's a possible
		// scenario in which a peer sends some messages and then crashes before writing its state to stable storage, so
		// when it comes back up, it reverts back to a lower msgCount. We should detect when this happens, and replay
		// the lost messages from another peer.
		if (localEntry.getPeerId() != originPeerId)
			throw new RuntimeException("processedIncomingMsg: peerid mismatch: " + localEntry.getPeerId() + " != " + originPeerId);
		if (localEntry.getMsgCount() + 1 > msgCount)
			throw new RuntimeException("processedIncomingMsg: msgCount for " + originPeerId + " went backwards: " + localEntry.getMsgCount() + 1 + " > " + msgCount);
		if (localEntry.getMsgCount() + 1 < msgCount)
			throw new RuntimeException("processedIncomingMsg: msgCount for " + originPeerId + " jumped forwards: " + localEntry.getMsgCount() + 1 + " < " + msgCount);

		localEntry.setMsgCount(msgCount);
		remoteEntry.setMsgCount(msgCount);
		this.localClockUpdate.recordUpdate(originPeerId, originIndex, msgCount);
	}

	/**
	 * Returns <code>true</code> if operations originating on the given peer ID are ready to be delivered to the
	 * application, and false if they need to be buffered. Operations are causally ready if all operations they may
	 * depend on (which had been processed by the time that operation was generated) have already been applied locally.
	 * We assume that pairwise communication between peers is totally ordered, i.e. that messages from one particular
	 * peer are received in the same order as they were sent.
	 */
	public boolean isCausallyReady(String remotePeerId) {
		Map<String, Long> localVclock = new HashMap<String, Long>();
		for (PeerVClockEntry entry : getPeerVClockList(PEER_INDEX_LOCAL)) {
			localVclock.put(entry.getPeerId(), entry.getMsgCount());
		}

		Map<String, Long> remoteVclock = new HashMap<String, Long>();
		for (PeerVClockEntry entry : getPeerVClockList(this.peerIdToIndex(remotePeerId))) {
			remoteVclock.put(entry.getPeerId(), entry.getMsgCount());
		}

		HashSet<String> allPeerIds = new HashSet<String>(localVclock.keySet());
		allPeerIds.addAll(remoteVclock.keySet());
		for (String peerId : allPeerIds) {
			Long local = localVclock.get(peerId);
			Long remote = remoteVclock.get(peerId);
			if ((!peerId.equals(remotePeerId)) && (local == null ? 0L : local) < (remote == null ? 0L : remote))
				return false;
		}
		return true;
	}

	/*
	 * Resets the tracking of messages received from other peers. This is done after a clock update has been broadcast
	 * to other peers, so that we only transmit a diff of changes to the clock since the last clock update.
	 */
	protected void resetClockUpdate() {
		this.localClockUpdate = new LocalClockUpdate();
	}

	public LocalClockUpdate getLocalClockUpdate() {
		return localClockUpdate;
	}

	@Override
	public String toString() {
		return "PeerMatrix [matrix=" + vectorClockMatrix + ", indexByPeerId=" + indexByPeerId + ", localClockUpdate=" + localClockUpdate + "]";
	}
}